### PR TITLE
*: Support stream agg.

### PR DESCRIPTION
**Pull request by hanfei1991**
1. move the logic of x operator push down to physical plan building phase.
2. change the cbo framework to support stream agg. It involves a concept of "group", e.g. group by a,b can be sorted by either a,b or b,a.
3. currently a few test still can't pass. I will fix it in next commit.

### CODE REVIEW

**Comment by hanfei1991** [comment]
@shenli @coocood @zimulala PTAL

File: executor/builder.go [hunk 1]
```diff
@@ -625,7 +550,6 @@ func (b *executorBuilder) buildTableScan(v *plan.PhysicalTableScan, s *plan.Sele
 	}
 	supportDesc := client.SupportRequestType(kv.ReqTypeSelect, kv.ReqSubTypeDesc)
 	if !memDB && client.SupportRequestType(kv.ReqTypeSelect, 0) {
```
Comment (shenli) [inline executor/builder.go]: This is already checked in plan builder.

Comment (hanfei1991) [inline executor/builder.go]: During plan building, this condition decides if we can push down the predicates.
During executor building, this condition decides to use which executor.

File: plan/plan.go [hunk 2]
```diff
@@ -142,7 +156,7 @@ type LogicalPlan interface {
 	// convert2PhysicalPlan converts logical plan to physical plan. The arg prop means the required sort property.
 	// This function returns two response. The first one is the best plan that matches the required property strictly.
 	// The second one is the best plan that needn't matches the required property.
-	convert2PhysicalPlan(prop requiredProperty) (*physicalPlanInfo, *physicalPlanInfo, uint64, error)
+	convert2PhysicalPlan(prop *requiredProperty) (*physicalPlanInfo, error)
```
Comment (shenli) [inline plan/plan.go]: Update comments.

File: executor/executor.go [hunk 3]
```diff
@@ -1061,7 +1061,14 @@ func (e *AggregationExec) Next() (*Row, error) {
...
-func (e *AggregationExec) getGroupKey(row *Row) ([]byte, error) {
+func (e *HashAggExec) getGroupKey(row *Row) ([]byte, error) {
+	if e.mode == plan.FinalAgg {
+		val, err := e.GroupByItems[0].Eval(row.Data, e.ctx)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		return val.GetBytes(), nil
+	}
 	if len(e.GroupByItems) == 0 {
```
Comment (shenli) [inline executor/executor.go]: if e.hasGby {}

File: executor/executor.go [hunk 4]
```diff
@@ -1113,6 +1120,102 @@ func (e *AggregationExec) innerNext() (ret bool, err error) {
...
+
+// Fields implements Executor Fields interface.
+func (e *StreamAggExec) Fields() []*ast.ResultField {
+	return e.ResultFields
+}
+
+// Next implements Executor Next interface.
+func (e *StreamAggExec) Next() (*Row, error) {
+	// In this stage we consider all data from src as a single group.
+	if !e.executed {
```
Comment (shenli) [inline executor/executor.go]: if e.executed {
  return nil, nil
}

File: executor/executor.go [hunk 5]
```diff
@@ -1113,6 +1120,102 @@ func (e *AggregationExec) innerNext() (ret bool, err error) {
...
+}
+
+// Fields implements Executor Fields interface.
+func (e *StreamAggExec) Fields() []*ast.ResultField {
+	return e.ResultFields
+}
+
+// Next implements Executor Next interface.
+func (e *StreamAggExec) Next() (*Row, error) {
+	// In this stage we consider all data from src as a single group.
```
Comment (shenli) [inline executor/executor.go]: The comment confuses me.

File: executor/executor.go [hunk 6]
```diff
@@ -1113,6 +1120,102 @@ func (e *AggregationExec) innerNext() (ret bool, err error) {
...
+				e.executed = true
+				break
+			}
+			groupKey, err = e.getGroupKey(row)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			var end bool
+			for _, af := range e.AggFuncs {
+				end, err = af.StreamUpdate(row.Data, groupKey, e.ctx)
```
Comment (shenli) [inline executor/executor.go]: check err

File: expression/aggregation.go [hunk 7]
```diff
@@ -126,6 +170,40 @@ func (af *aggFunction) getContext(groupKey []byte) *ast.AggEvaluateContext {
...
+	af.resultCtx = af.streamCtx
+	af.streamCtx = &ast.AggEvaluateContext{}
+	af.lastRowEncode = nil
+}
+
+func (af *aggFunction) getStreamedContext(groupKey []byte) (*ast.AggEvaluateContext, bool) {
+	end := false
+	if len(af.lastGroup) == 0 {
+		af.lastGroup = groupKey
+	} else if string(af.lastGroup) != string(groupKey) {
```
Comment (shenli) [inline expression/aggregation.go]: https://golang.org/pkg/bytes/#Compare Maybe better.

File: expression/aggregation.go [hunk 8]
```diff
@@ -31,17 +32,31 @@ type AggregationFunction interface {
...
+	// StreamUpdate updates data using streaming algo.
+	StreamUpdate(row []types.Datum, groupKey []byte, ctx context.Context) (bool, error)
+
+	NeedCount() bool
+
+	NeedValue() bool
+
+	SetMode(mode AggFunctionMode)
+
+	UpdateStreamResult()
```
Comment (shenli) [inline expression/aggregation.go]: Need comments for those newly added interfaces.

File: expression/aggregation.go [hunk 9]
```diff
@@ -126,6 +170,40 @@ func (af *aggFunction) getContext(groupKey []byte) *ast.AggEvaluateContext {
 	return ctx
 }
 
+func (af *aggFunction) checkDistinct(values ...types.Datum) (bool, error) {
+	bytes, err := codec.EncodeValue(nil, values...)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	if string(bytes) == string(af.lastRowEncode) {
```
Comment (shenli) [inline expression/aggregation.go]: bytes.Compare

File: executor/executor.go [hunk 10]
```diff
@@ -1111,6 +1118,101 @@ func (e *AggregationExec) innerNext() (ret bool, err error) {
...
+			end, err = af.StreamUpdate(row.Data, groupKey, e.ctx)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+		}
+		if end {
+			break
+		}
+	}
+	retRow := &Row{Data: make([]types.Datum, 0, len(e.AggFuncs))}
```
Comment (shenli) [inline executor/executor.go]: If there is groupby items and you do not get any row from e.src. Will this Next() return a row?

File: expression/aggregation.go [hunk 11]
```diff
@@ -190,6 +276,9 @@ func (cf *countFunction) Update(row []types.Datum, groupKey []byte, ectx context
 		if value.GetValue() == nil {
 			return nil
 		}
+		if cf.mode == FinalMode {
```
Comment (shenli) [inline expression/aggregation.go]: if cf.mode == FinalMode, len(cf.Args) will be 1?

Comment (hanfei1991) [inline expression/aggregation.go]: yes

File: executor/executor.go (hunk 10 above)
Comment (hanfei1991) [inline executor/executor.go]: yes, it will

File: plan/plan.go [hunk 12]
```diff
@@ -163,34 +176,50 @@ type PhysicalPlan interface {
 
 type baseLogicalPlan struct {
 	basePlan
-	propLen          int
-	sortedPlanInfo   *physicalPlanInfo
-	unSortedPlanInfo *physicalPlanInfo
-	count            uint64
+	planMap map[string]*physicalPlanInfo
+	proxy   Plan
```
Comment (shenli) [inline plan/plan.go]: What's this proxy used for?

Comment (hanfei1991) [inline plan/plan.go]: Let baseLogicalPlan implement "convert to physical plan" interface.

File: expression/aggregation.go [hunk 13]
```diff
@@ -157,6 +208,27 @@ func (af *aggFunction) updateSum(row []types.Datum, groupKey []byte, ectx contex
 	return nil
 }
 
+func (af *aggFunction) streamUpdateSum(row []types.Datum, groupKey []byte, ectx context.Context) (bool, error) {
+	ctx, end := af.getStreamedContext(groupKey)
+	a := af.Args[0]
+	value, err := a.Eval(row, ectx)
```
Comment (zimulala) [inline expression/aggregation.go]: Check this `err`.

File: expression/aggregation.go [hunk 14]
```diff
@@ -76,11 +85,22 @@ func NewAggFunction(funcType string, funcArgs []Expression, distinct bool) Aggre
...
+
 type aggFunction struct {
 	name         string
+	mode         AggFunctionMode
 	Args         []Expression
 	Distinct     bool
 	resultMapper aggCtxMapper
+	streamCtx    *ast.AggEvaluateContext
+	resultCtx    *ast.AggEvaluateContext
+	lastGroup    []byte
```
Comment (XuHuaiyu) [inline expression/aggregation.go]: Would currentGroup be better?

File: expression/aggregation.go (hunk 13 above)
Comment (XuHuaiyu) [inline expression/aggregation.go]: Condition like
``` go
if value.IsNull() {
      return end, nil
}
```
can be added like updateSum() does.

File: executor/union_scan.go [hunk 15]
```diff
@@ -100,10 +100,9 @@ type UnionScanExec struct {
 	Src   Executor
 	dirty *dirtyTable
 	// srcUsedIndex is the column offsets of the index which Src executor has used.
```
Comment (tiancaiamao) [inline executor/union_scan.go]: seems comment outdated? 
/s/srcUsedIndex/usedIndex

File: expression/aggregation.go [hunk 16]
```diff
@@ -203,28 +292,108 @@ func (cf *countFunction) Update(row []types.Datum, groupKey []byte, ectx context
...
+	if err != nil {
+		return errors.Trace(err)
+	}
+	ctx.Count += count.GetInt64()
+	return nil
+}
+
 // Update implements AggregationFunction interface.
 func (af *avgFunction) Update(row []types.Datum, groupKey []byte, ctx context.Context) error {
+	if af.mode == FinalMode && af.name == ast.AggFuncAvg {
```
Comment (XuHuaiyu) [inline expression/aggregation.go]: Why it is need to judge the function name here, would it be other names?

File: expression/aggregation.go [hunk 17]
```diff
@@ -31,17 +31,26 @@ type AggregationFunction interface {
 	// Update during executing.
 	Update(row []types.Datum, groupKey []byte, ctx context.Context) error
 
+	// StreamUpdate updates data using streaming algo.
+	StreamUpdate(row []types.Datum, groupKey []byte, ctx context.Context) (bool, error)
```
Comment (zimulala) [inline expression/aggregation.go]: I think this interface can be remove. Most of the operations of `Update` and `StreamUpdate` are the same. Only the number of return value is different.

Comment (hanfei1991) [inline expression/aggregation.go]: StreamUpdate calls getStreamContext and returns a bool value standing for whether the current group by item is equal to last group by items.

Comment (zimulala) [inline expression/aggregation.go]: Add a flag is enough.

Comment (hanfei1991) [inline expression/aggregation.go]: But how to get the bool value from Update() function to notice the outside world to get a row ?

File: expression/expression.go [hunk 18]
```diff
@@ -70,6 +70,7 @@ type Column struct {
 	DBName  model.CIStr
 	TblName model.CIStr
 	RetType *types.FieldType
+	ID      int64
```
Comment (shenli) [inline expression/expression.go]: What's this ID used for?

File: plan/expr_to_pb.go [hunk 19]
```diff
@@ -0,0 +1,331 @@
...
+	if column.Correlated {
+		return nil, nil
+	}
+
+	id := column.ID
+	// Zero Column ID is not a column from table, can not support for now.
+	if id == 0 {
+		return nil, nil
+	}
+	// its value is available to use.
```
Comment (shenli) [inline plan/expr_to_pb.go]: Can not understand the comment. When will column.ID be set to -1?

File: plan/logical_plan_builder.go [hunk 20]
```diff
@@ -393,6 +401,7 @@ func (b *planBuilder) buildLimit(src LogicalPlan, limit *ast.Limit) LogicalPlan
 		Count:           limit.Count,
 		baseLogicalPlan: newBaseLogicalPlan(Lim, b.allocator),
 	}
+	li.proxy = li
```
Comment (shenli) [inline plan/logical_plan_builder.go]: Seems the plans' proxy  always be theirself?

File: expression/aggregation.go (hunk 17 above)
Comment (zimulala) [inline expression/aggregation.go]: Change the return value of `Update` .

File: expression/aggregation.go [hunk 21]
```diff
@@ -166,11 +238,25 @@ func (sf *sumFunction) Update(row []types.Datum, groupKey []byte, ctx context.Co
...
+}
+
 // GetGroupResult implements AggregationFunction interface.
 func (sf *sumFunction) GetGroupResult(groupKey []byte) (d types.Datum) {
 	return sf.getContext(groupKey).Value
 }
 
+func (sf *sumFunction) GetStreamResult() (d types.Datum) {
+	defer func() {
+		sf.resultCtx = sf.streamCtx
```
Comment (zimulala) [inline expression/aggregation.go]: `GetStreamResult` only one return value so removing `defer` is better.

File: expression/aggregation.go [hunk 22]
```diff
@@ -350,7 +622,34 @@ func (ff *firstRowFunction) Update(row []types.Datum, groupKey []byte, ectx cont
...
+
 // GetGroupResult implements AggregationFunction interface.
-func (ff *firstRowFunction) GetGroupResult(groupKey []byte) (d types.Datum) {
+func (ff *firstRowFunction) GetGroupResult(groupKey []byte) types.Datum {
 	return ff.getContext(groupKey).Value
 }
+
+func (ff *firstRowFunction) GetStreamResult() types.Datum {
+	defer func() {
+		ff.resultCtx = ff.streamCtx
```
Comment (zimulala) [inline expression/aggregation.go]: The same as here.

File: expression/aggregation.go (hunk 21 above)
Comment (hanfei1991) [inline expression/aggregation.go]: you mean change it to 
`d = sf.resultCtx.Value
 sf.resultCtx = sf.streamCtx
return
`
like this ?

File: plan/match_property.go [hunk 23]
```diff
@@ -15,78 +15,90 @@ package plan
...
-	return &physicalPlanInfo{p: ts, cost: math.MaxFloat64}
+	return &physicalPlanInfo{p: ts, cost: math.MaxFloat64, count: infos[0].count}
 }
 
 // matchProperty implements PhysicalPlan matchProperty interface.
-func (is *PhysicalIndexScan) matchProperty(prop requiredProperty, rowCounts []uint64, _ ...*physicalPlanInfo) *physicalPlanInfo {
-	rowCount := float64(rowCounts[0])
+func (is *PhysicalIndexScan) matchProperty(prop *requiredProperty, infos ...*physicalPlanInfo) *physicalPlanInfo {
+	rowCount := float64(infos[0].count)
 	// currently index read from kv 2 times.
```
Comment (shenli) [inline plan/match_property.go]: Update this comment

File: plan/match_property.go [hunk 24]
```diff
@@ -15,78 +15,90 @@ package plan
...
-		}
-		if prop[matched].col.ColName.L != indexCol.Name.L {
-			if matched == 0 && i < is.accessEqualCount {
-				continue
+	matchedList := make([]bool, len(prop.props))
+	for i, idxCol := range is.Index.Columns {
+		findMatched := false
+		if matchedIdx < prop.sortKeyLen {
+			propCol := prop.props[matchedIdx]
+			if idxCol.Name.L == propCol.col.ColName.L {
```
Comment (shenli) [inline plan/match_property.go]: Consider idxCol.Length

File: plan/plan.go [hunk 25]
```diff
@@ -73,14 +76,6 @@ type Plan interface {
 	Fields() []*ast.ResultField
 	// SetFields sets the results fields of the plan.
 	SetFields(fields []*ast.ResultField)
-	// The cost before returning fhe first row.
-	StartupCost() float64
```
Comment (shenli) [inline plan/plan.go]: Why remove these interfaces?

Comment (hanfei1991) [inline plan/plan.go]: These will never be used

File: plan/plan.go [hunk 26]
```diff
@@ -163,34 +176,50 @@ type PhysicalPlan interface {
...
-		return nil, nil, 0
+func (p *baseLogicalPlan) getPlanInfo(prop *requiredProperty) (*physicalPlanInfo, error) {
+	key, err := prop.getHashKey()
+	if err != nil {
+		return nil, errors.Trace(err)
 	}
-	if len(prop) > p.propLen {
-		return nil, nil, 0
+	return p.planMap[string(key)], nil
+}
```
Comment (zimulala) [inline plan/plan.go]: Add a blank line between two functions.

File: expression/expression.go (hunk 18 above)
Comment (hanfei1991) [inline expression/expression.go]: ID is used to construct pb structure.

File: expression/aggregation.go (hunk 17 above)
Comment (hanfei1991) [inline expression/aggregation.go]: But currently we didn't apply stream alogrithm to distinct function, so you see there are the same logics in stream update and non-stream update.
If we implement stream distinct function, we will also write two different updates for them :)

File: expression/aggregation.go (hunk 21 above)
Comment (hanfei1991) [inline expression/aggregation.go]: but i think using defer is more clear.

File: plan/expr_to_pb.go (hunk 19 above)
Comment (hanfei1991) [inline plan/expr_to_pb.go]: It is ported from executor written by @zimulala . I guess she ported this from old plan. So maybe no one know why.

File: expression/aggregation.go (hunk 21 above)
Comment (zimulala) [inline expression/aggregation.go]: I think it's useless and `defer` for performance is not friendly.

Comment (hanfei1991) [inline expression/aggregation.go]: If we abandon this assignment, when we get nil from source executor, we can't get the correct ctx(we always need result ctx but stream ctx)

File: executor/executor.go [hunk 27]
```diff
@@ -1111,6 +1118,107 @@ func (e *AggregationExec) innerNext() (ret bool, err error) {
...
+	hasData      bool
+	mode         plan.AggregationType
+	ctx          context.Context
+	AggFuncs     []expression.AggregationFunction
+	GroupByItems []expression.Expression
+}
+
+// Close implements Executor Close interface.
+func (e *StreamAggExec) Close() error {
+	e.executed = false
```
Comment (tiancaiamao) [inline executor/executor.go]: `executed` need three kinds of state: not executed, executed, closed
we can't distinguish between not executed and closed if set `e.executed = false` here...

Comment (hanfei1991) [inline executor/executor.go]: closed is meaningless ....

Close() means "Initialize".

File: expression/aggregation.go [hunk 28]
```diff
@@ -76,11 +85,22 @@ func NewAggFunction(funcType string, funcArgs []Expression, distinct bool) Aggre
 
 type aggCtxMapper map[string]*ast.AggEvaluateContext
 
+type AggFunctionMode int
+
+const (
+	CompleteMode AggFunctionMode = iota
+	FinalMode
```
Comment (tiancaiamao) [inline expression/aggregation.go]: It's better to add code comments for AggreFunctionMode

File: expression/aggregation.go [hunk 29]
```diff
@@ -203,28 +292,108 @@ func (cf *countFunction) Update(row []types.Datum, groupKey []byte, ectx context
...
+}
+
 // GetGroupResult implements AggregationFunction interface.
 func (cf *countFunction) GetGroupResult(groupKey []byte) (d types.Datum) {
 	d.SetInt64(cf.getContext(groupKey).Count)
 	return d
 }
 
+func (cf *countFunction) GetStreamResult() (d types.Datum) {
+	defer func() {
```
Comment (tiancaiamao) [inline expression/aggregation.go]: avoid this kind of usage of defer, it's not go idiom

File: plan/expr_to_pb.go [hunk 30]
```diff
@@ -0,0 +1,331 @@
...
+				Children: []*tipb.Expr{pbExpr, v}}
+		}
+	}
+	return
+}
+
+func exprToPB(client kv.Client, expr expression.Expression) (*tipb.Expr, error) {
+	switch x := expr.(type) {
+	case *expression.Constant:
+		return datumToPBExpr(client, expr.(*expression.Constant).Value)
```
Comment (tiancaiamao) [inline plan/expr_to_pb.go]: expr.(*expression.Constant).Value -> x.Value ?

File: executor/executor.go [hunk 31]
```diff
@@ -1111,6 +1118,107 @@ func (e *AggregationExec) innerNext() (ret bool, err error) {
...
+func (e *StreamAggExec) getGroupKey(row *Row) ([]byte, error) {
+	if e.mode == plan.FinalAgg {
+		val, err := e.GroupByItems[0].Eval(row.Data, e.ctx)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		return val.GetBytes(), nil
+	}
+	if len(e.GroupByItems) == 0 {
+		return []byte{}, nil
```
Comment (zimulala) [inline executor/executor.go]: []byte{} -> nil
The same as line 1215.

File: plan/physical_plans.go [hunk 32]
```diff
@@ -108,6 +217,31 @@ type PhysicalHashSemiJoin struct {
...
+
+	Table     table.Table
+	Desc      bool
+	Condition expression.Expression
+}
+
+type AggregationType int
+
+const (
+	StreamedAgg AggregationType = iota
```
Comment (zimulala) [inline plan/physical_plans.go]: Add comments here.

File: expression/aggregation.go [hunk 33]
```diff
@@ -76,11 +85,24 @@ func NewAggFunction(funcType string, funcArgs []Expression, distinct bool) Aggre
...
+	FinalMode
+)
+
 type aggFunction struct {
 	name         string
+	mode         AggFunctionMode
 	Args         []Expression
 	Distinct     bool
 	resultMapper aggCtxMapper
+	streamCtx    *ast.AggEvaluateContext
```
Comment (zimulala) [inline expression/aggregation.go]: Add some comments for easy reading or change the variable name.

File: expression/aggregation.go [hunk 34]
```diff
@@ -126,6 +155,30 @@ func (af *aggFunction) getContext(groupKey []byte) *ast.AggEvaluateContext {
...
+	end := false
+	if len(af.curGroup) == 0 {
+		af.curGroup = groupKey
+	} else if bytes.Compare(af.curGroup, groupKey) != 0 {
+		af.curGroup = groupKey
+		end = true
+	}
+	if end {
+		af.resultCtx = af.streamCtx
+		af.streamCtx = &ast.AggEvaluateContext{}
```
Comment (zimulala) [inline expression/aggregation.go]: This line can put out of the `if` statement.

**Comment by hanfei1991** [comment]
@shenli @coocood @zimulala @tiancaiamao PTAL

**Pull request closed without merging**